### DISCUSSION TITLE
state-of-the-art windows virtualization in Ardour/JUCE and how spack could step in

### DISCUSSION THREAD

**Discussion by cosmicexplorer** [spack/spack#20779] [Ideas]
# Review of Wine Runtime Virtualization from Ardour

I began my journey with [this doc page](https://manual.ardour.org/working-with-plugins/windows-vst-support/) from the [Ardour audio software](https://ardour.org/). Upon reading it I was impressed to learn that Ardour supports a *very* tricky setup with [Wine](https://www.winehq.org/) so their users can seamlessly execute tools only ever released for windows.

## Wine Support Issues
But the project seems lukewarm on whether they'll keep it (any emphasis is my own):
> Thanks to the combined work of Torben Hohn, Kjetil Mattheusen, Paul Davis and a few other developers, it is possible to use Windows VST plugins (that is, plugins in VST format built and distributed for the Windows platforms) on Ardour running on Linux.
> 
> However, doing so has three **substantial downsides:**
> 
> 1. It requires a special build of Ardour that is fundamentally very different from normal builds
> 2. Support depends on Wine, a Windows "emulator"
> 3 As usual with plugins, a crashing plugin will take Ardour down with it—and crashes in Windows VST plugins are more likely when used in this way
> 
> **The dependence on Wine makes it almost impossible for the Ardour project to support this feature. Wine's functionality generally improves over time, but any given release of Wine may behave worse with some or all Windows VST plugins. It may even just crash Ardour completely.**
> 

## Interactions Between Virtualization Platforms at Once

They attempt to describe the internal difficulty of the current solution, and then compare alternatives:
> Step back and think about what "using Windows VSTs" really means: taking bits of software written with only one idea in mind—running on the Windows platform—and then trying to use them on an entirely different platform. It is a bit of a miracle (thanks largely to the incredible work done by the Wine project) that it works at all. But is this the basis of a stable, reliable DAW for a non-Windows platform? Getting Ardour on Linux to pretend that its really a Windows application running on Windows?
> 
> It is understandable that there are many outstanding plugins available as Windows VSTs and, that in many cases, no equivalent is available for Linux. If a workflow is so dependent on those plugins, Ardour should be used on Windows **(or potentially used with an actual Windows VST host running inside of Wine).** If the effort can be made, a better environment can be obtained by using a normal build of Ardour and *exploring the world of plugins built to run on Linux natively*. **This covers LADSPA, LV2 and Linux VST formats, and even some outstanding proprietary plugins such as those from Loomer.**
> 

From the above quoted sections, I have *(currently)* parsed out the subsequent remarks:
## People
- [ ] "Torben Hohn, Kjetil Mattheusen, Paul Davis and a few other developers" were *all* needed to make it possible to run this initial pass at windows VSTs in Wine.
    - [ ] They sound like possibly good contacts to make.

## Runtime Virtualization Options Addressed
*Thinking about the **interactions between these shorter-lived and composable virtualization mechanisms** reminds me of things @vsoch has said or I've inferred from the words in spack/build-si-modeling#1:*
- [ ] Containers have a `uri` separate from the checksum to distinguish them.
- [ ] Containers may write to a file within its accessible premade Dockerfile/etc fs, or into the parent fs, at any time.
- [ ] Containers may be a way to directly implement checksummed, cacheable process executions depending on runtime: https://github.com/bazelbuild/remote-apis/blob/9e72daff42c941baaf43a4c370e2607a984c58a7/build/bazel/remote/execution/v2/remote_execution.proto#L373-L394
- [ ] Containers may spawn processes which may be host executables or other containers.

### Current Best and Alternatives Considered
- [ ] It is now possible to use Windows VST plugins (that is, plugins in VST format built and distributed for the Windows platforms) on Ardour running on Linux.
- [ ] It is *maybe* also possible to use Windows VST plugins on Ardour for Windows, running a Windows VST Host in Wine.

## Build Issues with Wine
There were several times the `wine` tool was challenged:

- [ ] **Wine is not to be trusted at build time or runtime!** *(this came as a huge surprise to me!)*:
    - [ ] What special build process does Ardour require from wine?
    - [ ] What kind of variants and constraints does this build process consume/produce? *See #20523!*
- [ ] When and who chooses to perform a subsequent update of wine in this model?
    - [ ] Is the user made aware of it?
    - [ ] Can the user manually cancel the update easily?
    - [ ] Where is wine resolved from?
- [ ] What virtualization (e.g. #20359) or isolation (e.g. https://github.com/spack/spack/issues/20260) processes does Wine make use of?
    - [ ] Have they written code we can use to address any of the pain points remaining in https://github.com/spack/spack/issues/20260?
    - [ ] They propose
- [ ] Which underlying mechanism leads to Wine's nondeterministic failures with Ardour on Linux?
    - [ ] Who on the Wine team owns this?

## Controlling the Build is Easier than the Halting Problem
- [ ] *Exploring the world of plugins built to run on Linux natively* indicates that Ardour considers it *much* more maintainable **to control this multi-platform build system instead of a runtime virtualization scheme**.
    - [ ] **I think @vsoch might have had some very useful thoughts on spack/build-si-modeling#1 about this.**
    - [ ] In the case that a spack install script is executing, and a nested subprocess `PATH` lookup fails (cc @tgamblin), we can solve the problem by either:
        - [ ] Install the Spack package(s) which provide that *path entry*.
            - [ ] **We do not have any introspection methods defined for the more granular depencies of #20523 at all!**
        - [ ] Install using an OS package manager, and ensure `spack external find` picks it up.
    - [ ] Install dependencies _into_ a container, then execute the container.

# Possible Simbiosis for Spack in Support

## Multiple Competing Standards
*Concluding the extended quote from above:*
> **A Plea To Plugin Manufacturers**
> 
> Please consider porting your plugins so that users can enjoy them on Linux too. Several other commercial plugin developers have already done this. You can choose between using "Linux VST" (which is what Loomer and others have done)—you will find **toolkits like JUCE** that help to make this fairly easy—or **using LV2 format** *which is ultimately more flexible but probably requires more work*. **We have users—thousands of users—on Linux who would like to use your plugins.**

Ok, so there are multiple competing standards:
- [Windows VST](https://help.ableton.com/hc/en-us/articles/209071729-Using-VST-plug-ins-on-Windows) via emulation.
- [Linux VST](http://linux-sound.org/linux-vst-plugins.html)
    - *TODO: How does this differ from Windows VST?*
- [Linux LV2](https://lv2plug.in/)
- Along with room for helper toolkits like [JUCE](https://juce.com)!

These are all purely for realtime audio production, so are less likely to be directly useful. However, one very powerful assumption *is* shared with the spack model, which is that **users and especially developers of the build tool do not have control over updating how sources are changed**.

However, who are these users again? That's right, mostly Linux users as Ardour was Linux-only for many many years. And again, they're described like this:
> **We have users—thousands of users—on Linux**

It seems like attracting thousands of users who care so deeply about using Linux they will forsake other audio software could be a really good thing for spack to attract.

## JUCE is "Cross-Platform"
See https://docs.juce.com/master/tutorial_manage_projucer_project.html:
> the Projucer is not just about that — it is also a very powerful cross-platform tool to manage all settings for your project, add source files, and much more. As you develop your JUCE application, the Projucer will be your faithful companion throughout the process. This tutorial gives you an overview over some essential features.

Ok, so JUCE *says* it does cross-platform builds, but what does it *generate*? Well, two things I can easily tell we care about:
1. They maintain generators for windows, OSX, and Linux **IDE project definitions.**
    - [ ] If spack doesn't already have a huge amount of work done on this I can't see, it might be super super useful to depend as directly as possible on the work JUCE has done for generating project definition files.
    - [ ] In particular, this is useful on Windows (I believe), because that also includes ways to invoke Visual Studio which spack currently doesn't know anything about.

![steal-windows-build](https://user-images.githubusercontent.com/1305167/104205681-94fc8680-5426-11eb-91c8-7f2d83440b43.png)

2. The maintain a concept of "modules". It's not yet clear what these are, but they appear to be automatically inferred sub-project boundaries within a larger parent repo, at the very least.
    - [ ] Determine whether the concept is the same or in any way compatible with existing spack "modules".
    - [ ] Being able to infer these project boundaries alone might be useful for spack in being able to produce a quick overview of what's currently in the repo.
    - [ ] Note that modules seem to exist as _subsets_ of a larger build in the below image, but it's possible they're somehow independent of it.

![improve-their-modules](https://user-images.githubusercontent.com/1305167/104205668-9168ff80-5426-11eb-9bd1-da9a7f585f7c.png)

**Comment by vsoch** [comment]
To step back a second - the goal of this exploration is to provide audio software to users (some that is only available on Windows?) - the first is Ardour that requires wine / containers, and the second is JUCE which is already cross platform? And we would extend this idea to say if we can figure out an approach for these specific software, it can be extended to others? Is there a user / issue that is requesting these packages? It's definitely an interesting idea, although I never considered spack to provide these kind of (non Linux) tools.

For containers, getting audio and video to work with Docker is always a hassle. But wine, audio, and video/display work very well with Singularity, I've built several Windows games into a Singularity container and have it run fairly well. [Here is an example](https://github.com/bbodi/rustarok/tree/master/docker) of a game in rust (and the [Singularity.wine](https://github.com/bbodi/rustarok/blob/master/docker/Singularity.wine) file), [an example scientific application](https://github.com/deisseroth-lab/two-photon) that is first built with docker and converted to Singularity, and a [third example](https://gist.github.com/vsoch/1edb43e517e79172bded581a5a61b53a) that provides more of an interactive installer. The issue, of course, is that a build requires sudo as it's from a definition file, and I suspect most centers won't install Docker where spack is. The "non sudo" approach is to provide the pre-built container, and have users just pull it to "install" the tool. Which brings me to a question - if spack is a package manager, is there some future where a container URI could just be pulled? Is there some kind of potential future for a spack container repository with everything pre-built and ready to go? Could a user define a package that is just a URI, and then have the install sequence be to pull it and put it somewhere?

Most approaches with wine might be complicated to run via spack because ideally you want to bind the users's `.wine` folder in their home (does spack have a current strategy for handling user specific configuration?). The whole process also seems a bit messy - wine sometimes is just buggy. I'm not sure how well it would work. So I'd say - it would be good to first clarify the goals of being able to do something like this (short and long), and if we decide those are important, we can discuss more of the above.
